### PR TITLE
add basic overview of when to use type declarations

### PR DESCRIPTION
**Pull request by stevengj**
Questions about when to declare function arguments come up over and over on discourse, and so it seems useful to put a basic summary early in the chapter on Functions, rather than deferring it to later chapters.

### CODE REVIEW

**Review by timholy — APPROVED** [review]
This seems helpful and in the right place. There are other reasons to add types, but these are the most common and most relevant for newcomers.

**Comment by timholy** [comment]
Maybe slightly less emphasis, though? Feels a bit like shouting, though I understand why these points are worth emphasizing.

File: doc/src/manual/functions.md [hunk 1]
```diff
@@ -63,6 +63,23 @@ are identical to the passed values. Modifications to mutable values (such as `Ar
...
+* **Clarity:** Type declarations can serve as a form of documentation, indicating what argument types were expected by the programmer.
+
+However, it is a **common mistake to overly restrict the argument types**, which can unnecessarily limit the applicability of the function and prevent it from being re-used in circumstances you did not anticipate.    For example, the `fib(n::Integer)` function above works equally well for `Int` arguments (machine integers) and `BigInt` arbitrary-precision integers (see [BigFloats and BigInts](@ref)), which is especially useful because Fibonacci numbers grow exponentially rapidly and will quickly overflow any fixed-precision type like `Int` (see [Overflow behavior](@ref)).  If we had declared our function as `fib(n::Int)`, however, the application to `BigFloat` would have been prevented for no reason.   In general, you should use the **most general applicable abstract types** for arguments, and **when in doubt, omit the argument type**.  You can always add argument-type specifications later if they become necessary, and you *don't sacrifice performance or functionality* by omitting them.
```
Comment (lungben) [inline doc/src/manual/functions.md:81]: however, the application to `BigFloat` 
This should be `BigInt`, right?
Otherwise, very helpful!

Comment (stevengj) [reply doc/src/manual/functions.md:81]: Right, will fix.

**Comment by stevengj** [comment]
With regard to bold/emphasized text, I was influenced a lot by Jakob Nielsen's usability studies, which suggests that online readers typically scan text rather than reading from top to bottom, and so it is useful to [highlight key ideas in bold](https://www.nngroup.com/articles/how-users-read-on-the-web/) to aid scannability.

However, I can certainly cut down on this if people prefer.

**Comment by JeffBezanson** [comment]
Nice!

**Pull request merged**